SR-14635: __SwiftValue should be transparent to casting

tbkka opened the pull request:
This PR changes the casting machinery to avoid casting `__SwiftValue` boxes
directly.  This forces the caster to instead unwrap `__SwiftValue` boxes and
retry with the inner content.  This results in boxed values being cast like the
inner content.

This fixes the behavior in situations like the following:
```
   let t = ...
   let s = t as Any as! AnyObject
   // `s` is now a `__SwiftValue` box
   // Next line should be true iff t conforms to NSCopying
   // Prior to this change, it always succeeds
   s is NSCopying
```

After this change, the above cast succeeds only if `t` actually
conforms to `NSCopying`.

This is a follow-on to PR #37683.
Related to: SR-14635

tbkka committed:
SR-14635: __SwiftValue should be transparent to casting

This PR changes the casting machinery to avoid casting `__SwiftValue` boxes
directly.  This forces the caster to instead unwrap `__SwiftValue` boxes and
retry with the inner content.  This results in boxed values being cast like the
inner content.

This fixes the behavior in situations like the following:
```
   let t = ...
   let s = t as Any as! AnyObject
   // `s` is now a `__SwiftValue` box
   // Next line should be true iff t conforms to NSCopying
   // Prior to this change, it always succeeds
   s is NSCopying
```

After this change, the above cast succeeds only if `t` actually
conforms to `NSCopying`.

This is a follow-on to PR#37683.
Related to: SR-14635

tbkka commented:
@swift-ci Please test

tbkka committed:
Fix calling convention for unboxFromSwiftValueWithType

tbkka commented:
@swift-ci Please test

mikeash commented on stdlib/public/runtime/DynamicCast.cpp:
```diff
@@ -425,6 +425,30 @@ tryCastUnwrappingObjCSwiftValueSource(
+      destLocation, destType,
+      const_cast<OpaqueValue *>(srcInnerValue), srcInnerType,
+      destFailureType, srcFailureType,
+      /*takeOnSuccess=*/ false, mayDeferChecks);
```
I don't think this is right. If `swift_unboxFromSwiftValueWithType` succeeds, then we return success but nothing places the value into `destLocation`. If it fails, then we use the uninitialized `srcInnerValue` with `tryCast`. The third parameter is the expected type of the inner value, so `&srcInnerType` isn't going to do what we want. I think the right approach here would be to either grab `value` directly from the `__SwiftValue` object by offset, or call `swift_unboxFromSwiftValueWithType` with `Any` as the destination type, then continue casting with the output value (and, presumably, fail loudly if that unboxing somehow fails).

tbkka replied:
Yes, I did a bad copy/paste job here.  Looking at it more carefully, it looks like `swift_unboxFromSwiftValueWithType` is really just a casting operation itself, so invoking it to recursively perform the cast is probably the easiest approach.  Let's see what CI thinks of that.

mikeash replied:
When I wrote my first comment, I somehow didn't realize that all the `as? T` expressions in the implementation of `swift_unboxFromSwiftValueWithType` will actually result in full-fledged casting, and I was thinking that it would have to be an exact match. This new version is nice!

tbkka committed:
Use SwiftValue unpacking API correctly

tbkka commented:
@swift-ci Please test

al45tair reviewed (approved):
Looks good to me. Nice to have a fix for the unexpected NSCopying conformance.

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please test macOS Platform

tbkka committed:
Provide a Bincompat override for the newly tightened casting semantics

tbkka commented:
Rebased on top of current `main` to simplify testing.

tbkka commented:
Added a bincompat hook to allow selectively enabling the newly tightened casting semantics

tbkka committed:
Flip the bincompat switches on the tightened casting semantics
so the new behavior is always enabled by default on all OSes.

(This may change depending on whether anyone is actually depending
on this behavior.)

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please test macOS

tbkka commented:
Apparently, I'm missing some tests; when I flip the bincompat switch here, everything still passes.  Hmmmm.....

tbkka committed:
Enable the tests for this particular cast when running against a sufficiently-new runtime

tbkka commented:
I found the missing test (it was there but just disabled).  Enabled it and now we should be in pretty good shape.

tbkka commented:
@swift-ci Please test

tbkka commented:
Only one failure this time, apparently unrelated.

tbkka commented:
@swift-ci Please smoke test macOS platform
